nighthawk/test/mocks.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <chrono>
+#include <memory>
+
+#include "gmock/gmock.h"
+
+#include "test/test_common/simulated_time_system.h"
+
+#include "nighthawk/common/platform_util.h"
+#include "nighthawk/common/rate_limiter.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+
+constexpr std::chrono::milliseconds TimeResolution = 10ms;
+
+class SimulatedTimeAwarePlatformUtil : public PlatformUtil {
+public:
+  SimulatedTimeAwarePlatformUtil() : time_system_(nullptr) {}
+
+  void yieldCurrentThread() const override {
+    ASSERT(time_system_ != nullptr);
+    time_system_->setMonotonicTime(time_system_->monotonicTime() + TimeResolution);
+  }
+  void setTimeSystem(Envoy::Event::SimulatedTimeSystem& time_system) {
+    time_system_ = &time_system;
+  }
+
+private:
+  Envoy::Event::SimulatedTimeSystem* time_system_;
+};
+
+class MockPlatformUtil : public PlatformUtil {
+public:
+  MockPlatformUtil() { DelegateToSimulatedTimeAwarePlatformUtil(); }
+
+  MOCK_CONST_METHOD0(yieldCurrentThread, void());
+
+  void setTimeSystem(Envoy::Event::SimulatedTimeSystem& time_system) {
+    simulated_time_aware_platform_util_.setTimeSystem(time_system);
+  }
+
+private:
+  void DelegateToSimulatedTimeAwarePlatformUtil() {
+    // When this is called we are in a tight spin loop. SimulatedTimeAwarePlatformUtil moves the
+    // simulated time forward, avoiding the tests hanging.
+    ON_CALL(*this, yieldCurrentThread())
+        .WillByDefault(testing::Invoke(&simulated_time_aware_platform_util_,
+                                       &SimulatedTimeAwarePlatformUtil::yieldCurrentThread));
+  }
+
+  SimulatedTimeAwarePlatformUtil simulated_time_aware_platform_util_;
```

Comment:
Why the extra level of delegation?

Reply:
See the comment above?:
```
    // When this is called we are in a tight spin loop. SimulatedTimeAwarePlatformUtil moves the
    // simulated time forward, avoiding the tests hanging.
```

Reply:
Why not just inject `SimulatedTimeAwarePlatform` directly?

Reply:
I'm not 100% sure I understand.. I rewrote this a bit, please check. I would like to be able to `EXPECT_CALL` on this, we use that in some of the sequencer tests:

```
  // We will be stepping in time at precisely the rate limiter frequency below.
  // As the callbacks won't complete, the sequencer will never consider itself idle.
  // Hence, no spinning, and no calls to yieldCurrentThread.
  EXPECT_CALL(platform_util_, yieldCurrentThread()).Times(0);
```

If there is a better way, can you post an example?

Reply:
I think it would be best to separate the mock from the fake (simulated) implementation. You can still have individual test cases use the mock, and instantiate it to delegate as needed, it just feels a bit icky baking in one particular delegated behavior into the mock.

Mocks should generally be lightweight and policy free, although in Envoy we violate this quite a bit and do some lightweight simple behaviors; generally not fully delegation to a different class though.

Reply:
I'm willing to drop this one for now, looking at the code again, I feel that this provides the required test capabilities and we can iterate later on the structure as we get more use cases for the mock time.